lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,275 @@
+extract_documentation([{attribute, _Anno, doc, Doc}=_AST | T], State) ->
+    extract_documentation(T, update_doc(State, Doc));
+extract_documentation([AST0 | _T]=AST,
+                      #docs{meta = #{ equiv := {call,_,{atom,_,EquivF},Args}} = Meta}=State)
+    when is_tuple(AST0) andalso (tuple_size(AST0) > 2 orelse tuple_size(AST0) < 6) ->
+    Meta1 = Meta#{ equiv := {EquivF, length(Args)}},
+    extract_documentation(AST, update_meta(State, Meta1));
+extract_documentation([AST0 | _T]=AST,
+                      #docs{ meta = #{ equiv := {EquivF,EquivA}}=Meta}=State)
+  when is_tuple(AST0) andalso (tuple_size(AST0) > 2 andalso tuple_size(AST0) < 5) andalso
+       (element(3, AST0) =:= function orelse
+        element(3, AST0) =:= type     orelse
+        element(3, AST0) =:= opaque   orelse
+        element(3, AST0) =:= callback) ->
+    Kind = element(3, AST0),
+    Doc = io_lib:format("Equivalent to `~ts~p/~p`",[prefix(Kind), EquivF,EquivA]),
+    State1 = State#docs{ meta = maps:remove(equiv, Meta) },
+    extract_documentation(AST, update_doc(State1, Doc));
+extract_documentation([{function, _Anno, _F, _A, _}=AST0 | _]=AST,
+                      #docs{ meta = #{ equiv := {EquivF,EquivA}}=Meta}=State) ->
+    Kind = element(1, AST0),
+    Doc = io_lib:format("Equivalent to `~ts~p/~p`",[prefix(Kind), EquivF,EquivA]),
+    State1 = State#docs{ meta = maps:remove(equiv, Meta) },
+    extract_documentation(AST, update_doc(State1, Doc));
+extract_documentation([{function, Anno, F, A, [{clause, _, ClauseArgs, _, _}]}=_AST | T], State) ->
+    FunDoc = template_gen_doc({function, Anno, F, A, ClauseArgs}, State),
+    [FunDoc | extract_documentation(T, reset_state(State))];
+extract_documentation([{function, Anno, F, A, _Body}=_AST | T],
+                      #docs{doc = Doc}=State) when Doc =/= none ->
+    {Slogan, DocsWithoutSlogan} =
+        %% First we check if there is a doc prototype
+        case extract_slogan(Doc, F, A) of
+            undefined -> {io_lib:format("~p/~p",[F,A]), Doc};
+            SloganDocs -> SloganDocs
+        end,
+    AttrBody = {function, F, A},
+    FunDoc = gen_doc(Anno, AttrBody, Slogan, DocsWithoutSlogan, State),
+    [FunDoc | extract_documentation(T, reset_state(State))];
+extract_documentation([{attribute, Anno, TypeOrOpaque, {Type, _, TypeArgs}}=_AST | T], State)
+  when TypeOrOpaque =:= type orelse TypeOrOpaque =:= opaque ->
+    Args = fun_to_varargs(TypeArgs),
+    FunDoc = template_gen_doc({type, Anno, Type, length(Args), Args}, State),
+    [FunDoc | extract_documentation(T, reset_state(State))];
+extract_documentation([{attribute, Anno, callback, {{CB, A}, [Fun]}}=_AST | T], State) ->
+    Args = fun_to_varargs(Fun),
+    FunDoc = template_gen_doc({callback, Anno, CB, A, Args}, State),
+    [FunDoc | extract_documentation(T, reset_state(State))];
+extract_documentation([_H|T], State) ->
+    extract_documentation(T, reset_state(State));
```

Comment:
The state should not be reset here, as if we have the following code:

```
-doc "hello".
-spec foo() -> atom().
foo() -> ok.
```

The documentation will not be generated.

Reply:
Thanks, good catch on the ordering of things.
95% test coverage does not mean anything if all possible tests cases are not added :)

Reply:
We cannot catch all mistakes. Hopefully in the scenario above the user till have put a `-doc` before main as well and then will get a warning for that.